### ISSUE TITLE
Request to allow transcode-video to continue with default crop values if both tests differ when using detect-crop

### ISSUE THREAD

**Issue by cameronks** [donmelton/video_transcoding#56]
When using detect-crop in the same argument as transcode-video, if the results differ between mplayer and handbrake, the script will fail. I'd like to request an option to allow the script to continue with default 0:0:0:0 cropping values instead of the script halting.

The is beneficial if you want to _attempt_ to get correct cropping values, but trust if results differ between the two tests, something must be up, so continue safely with default values.

This is even more beneficial when you load up multiple inputs (like DVD extras or multiple movies) and you want to attempt to crop correctly when the tests agree, but don't want to monitor it.

**Comment by donmelton** [comment]
@cameronks Wow! We were talking about this on Twitter just minutes ago. Thanks for filing the issue request so quickly!

I was actually thinking about this a bit after our tweet storm, and maybe instead of changing the current behavior of `--crop detect`, I could add a new argument. Maybe something like `--crop detect-fallback` or `--crop detect-safe`?

Would that work for you?

**Comment by cameronks** [comment]
Actually that's exactly what I envisioned. --crop detect should stay the way it is, but an alternate like that would be great!

**Comment by donmelton** [comment]
@cameronks Excellent! OK, I'll put that on the TODO list for 0.8.

And I'll report back here by next week on the feasibility once (I'm sober and) I take a look at the code again.

**Comment by donmelton** [comment]
@cameronks It seems I sobered up enough to look at the code tonight. I don't think this is difficult and I can even see two ways to code it. So, the hard part now is to 1) pick the right name for the option argument and 2) think ahead as to whether we may need more argument names in the future.

I can also see (now that I think about it more) an alternate way to do this which is simply to have another option like `--fallback-crop T:B:L:R` where you could actually specify whether you wanted `0:0:0:0` or another set of values on detection fail. Any opinions on that idea?

**Comment by donmelton** [comment]
@cameronks OK, I have an even better idea for that new option! Instead of specifying a `T:B:L:R` bounds as a fallback (which, when I think about, almost no one would want to do), how about this instead:

```
--fallback-crop none|auto
```
...where you could choose `0:0:0:0` (your original idea) or the automatic behavior from HandBrake?

**Comment by cameronks** [comment]
The name sounds good, and I agree, I can't think of a use case to pick a specific fallback crop. "None" to me is the safe zone when the tests conflict. 

I could see a use case for going with the default handbrake settings too, but I'm my experience, Handbrake gets this wrong too many times, especially on sources like DVD extras where the aspect ratio changes quite a bit. But I like it, as it's the opposite of "None". And now there would be three options for the user if the crop tests conflict: 1) Stop, 2) Continue with no crop, 3) Continue with Handbrake crop.

**Comment by donmelton** [comment]
@cameronks Good point about HandBrake getting the crop wrong often enough to be worrisome.

Of course, if you're using `--crop detect` on a disc image directory instead of a single file, then you never fail because HandBrake data is all you can get. Crop information from MPlayer is unavailable for disc image directory input.

Also, falling back to the automatic crop of HandBrake is probably a bad idea. I should have thought of that. You really want the constrained version of that crop data. And it would be nice to have the choice of using the constrained crop data from MPlayer if it's available.

So, how about this (to add another argument option into the mix)?

```
--fallback-crop handbrake|mplayer|none
```

And, looking at the code now, this wouldn't be very hard to implement.

**Comment by donmelton** [comment]
@cameronks Here's a crazy idea: add another option to control whether `--detect crop` constrains the crop values. That's already available in the `detect-crop` tool itself.

Here's how the help text would look for all this:

```
    --crop T:B:L:R  set video crop values (default: 0:0:0:0)
                      (use `--crop detect` for optimal crop values)
                      (use `--crop auto` for `HandBrakeCLI` behavior)
    --no-constrain-crop
                    don't constrain `--crop detect` to optimal shape
    --fallback-crop handbrake|mplayer|none
                    select fallback crop values if `--crop detect` fails
```

What do you think?

**Comment by cameronks** [comment]
I'm actually not clear on the definition of constrained crop. Is it the lesser (more safe) of the values between handbrake and mplayer?

**Comment by donmelton** [comment]
@cameronks The detected crop is always the lesser of the values from HandBrake and MPlayer which makes it the most safe. "Constraining" means the algorithm to determine optimal shape always crops from the top and bottom or from the left and right, never from both axes.

**Comment by cameronks** [comment]
Ah, I see, so --no-constrain-crop would allow the video to be cropped from both axes using the info from --crop detect?

**Comment by donmelton** [comment]
@cameronks Yes, which is exactly what the `detect-crop` tool does when you pass it the `--no-constrain` option.

**Comment by cameronks** [comment]
Well, that looks like it covers all options then. I love it!

**Comment by donmelton** [comment]
@cameronks Excellent! OK, this will definitely go into version 0.8 then. Which will probably be out in the next few weeks. (I'm trying to get some additions and changes to the audio API completed as well which is why it will take so long.)

**Comment by cnrd** [comment]
May I propose yet another option for --fallback-crop?

`--fallback-crop ask`

where the script would pause and wait for the user to make a decision about the cropping, the user would then be able to supply the cropping in the T:B:L:R format.

This would be nice in the case of batch conversion, where --crop detect can be used, and in the case of a problem, the conversion will stop and wait fo the users decision on cropping, the reason for this is that I'm sometimes converting old movies where the source used for DVD is tape which makes the edges very uneven, which always screws up automatic detection.

**Comment by donmelton** [comment]
@cnrd I'll consider that. But such a feature would be the only case where any of the tools in the package behaved interactively. And if I Implemented the option to ask, I'm not sure it would be for such complicated input. Maybe just a choice between HandBrake and MPlayer.

**Comment by mrverrall** [comment]
I don't think 'ask' makes sense. I see the new options as tools to improve the automation of transcoding. If I'm available for interactive input I'll just use detect-crop and input the values manually.

**Comment by donmelton** [comment]
@mrverrall Agreed. Interactive input is the wrong approach for a tool that's typically run in batch mode.

**Comment by donmelton** [comment]
@cameronks OK, I have a patch working for this and I'm testing it now. Even easier than I thought. Stay tuned...

**Comment by donmelton** [comment]
@cameronks OK, the new options have been released in version 0.8.0. Just update to get it.

Thanks again!
